pandas/core/dtypes/concat.py

```diff
@@ -45,6 +46,8 @@ def get_dtype_kinds(l):
             # if to_concat contains different tz,
             # the result must be object dtype
             typ = str(arr.dtype)
+        elif is_range(arr):
```

Comment:
``Index._concat`` has no idea of the specific types of indexes, and rightly so... it [uses](https://github.com/pandas-dev/pandas/pull/16236/files#diff-248241e14f081931cc805f9d0e137a5fR1748) ``get_dtype_kinds`` just to test whether different types of indexes are being concatenated.